test/mako/alerter/github/issue.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/go-github/github"
+	"knative.dev/pkg/test/mako/alerter"
+
+	"knative.dev/test-infra/shared/ghutil"
+)
+
+const (
+	// perfLabel is the Github issue label used for querying all auto-generated performance issues.
+	perfLabel       = "auto:perf"
+	daysConsiderOld = 10 // arbitrary number of days for an issue to be considered old
+
+	// issueTitleTemplate is a template for issue title
+	issueTitleTemplate = "[performance] %s"
+
+	// issueBodyTemplate is a template for issue body
+	issueBodyTemplate = `
+### Auto-generated issue tracking performance regression
+* **Test name**: %s`
+
+	// reopenIssueCommentTemplate is a template for the comment of an issue that is reopened
+	reopenIssueCommentTemplate = `
+New regression has been detected, reopening this issue:
+%s`
+
+	// newIssueCommentTemplate is a template for the comment of an issue that has been quiet for a long time
+	newIssueCommentTemplate = `
+A new regression for this test has been detected:
+%s`
+
+	// closeIssueComment is the comment of an issue when we close it
+	closeIssueComment = `
+The performance regression goes way for this test, closing this issue.`
+)
+
+// IssueOperations defines operations that can be done to github
+type IssueOperations interface {
```

Comment:
Do we need the interface?

Reply:
It's required if we want the struct to be private, people will have to use the `Setup` function to construct `issueHandler`, in which we enforce the validations. It's the same factory pattern as what I used before - https://github.com/knative/pkg/blob/366ab85660d4e1c260795bd23c2aa09a4721c282/test/vegeta/pacers/combined_pacer.go#L48

Reply:
Yeah but you can just return the struct. Interfaces are more of a client side construct in Go (vs Java).
So unless you are sure to have multiple implementations, just return the struct.
E.g.: https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8